openhtf/plugs/uart_device.py

```diff
@@ -0,0 +1,134 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""SerialPlug used for serial communication
+Requires the following keys in .yaml configuration file AND declared.
+
+.yaml file
+  dut_interface: serial
+  serial_port:
+  serial_baudrate:
+  serial_parity:
+  serial_stopbits:
+  serial_bytesize:
+  serial_rtscts:
+  serial_timeout:
+
+declared in your project
+  conf.Declare('dut_interface', 'specify interface type: serial or adb for now')
+  conf.Declare('serial_port', 'port name of serial device. OS dependent')
+  conf.Declare('serial_baudrate')
+  conf.Declare('serial_parity')
+  conf.Declare('serial_stopbits')
+  conf.Declare('serial_bytesize')
+  conf.Declare('serial_rtscts')
+  conf.Declare('serial_timeout')
+"""
+
+import time
+import serial
+import binascii
+import glob
+import openhtf.plugs as plugs
+import openhtf.conf as conf
+import logging
+
+class SerialPlug(plugs.BasePlug):
+  '''A Plug for serial communication.'''
+
+  @conf.InjectPositionalArgs
+  def __init__(self):
+    logging.info('Instantiating %s with these settings:' % (type(self).__name__))
+    logging.info(' port: %s' %conf.Config().serial_port)
+    logging.info(' baudrate: %s' %conf.Config().serial_baudrate)
+    logging.info(' parity: %s' %conf.Config().serial_parity)
+    logging.info(' stopbits: %s' %conf.Config().serial_stopbits)
+    logging.info(' bytesize: %s' %conf.Config().serial_bytesize)
+    logging.info(' rtscts: %s' %conf.Config().serial_rtscts)
+    logging.info(' timeout: %s' %conf.Config().serial_timeout)
+
+    self._portHandler = None
+    self._portOpened = False
+
+  def OpenPort(self):
+    logging.info('%s: %s !' % (type(self).__name__, self.OpenPort.__name__))
+    # get exact port name for DUT
+    ports = glob.glob(conf.Config().serial_port)
+
+    if (len(ports) != 1):
+      logging.info('there are %d!'%(len(ports)))
+    else:
+      logging.info('got: %s' %ports[0])
+
+    try:
+      # configure the serial connections (the parameters differs on the device you are connecting to)
+      self._portHandler = serial.Serial(
+        port=ports[0],
+        baudrate=int(conf.Config().serial_baudrate),
+        parity=conf.Config().serial_parity,
+        stopbits=int(conf.Config().serial_stopbits),
+        bytesize=int(conf.Config().serial_bytesize),
+        rtscts=bool(int(conf.Config().serial_rtscts))
+      )
+      logging.info('%s open success!' %ports[0])
+      self._portOpened = True
+      return True
+    except (OSError, serial.SerialException):
+      logging.info("port open failed.")
+      return False
+
+  def SendRawCmd(self, cmd):
+    # convert ascii cmd to raw bytes, send over uart, and convert raw byte results back to ascii for display
+    logging.info('serial input: ' + cmd)
+    inputbinary = binascii.unhexlify(cmd)
+    self._portHandler.write(inputbinary)
+    out = ''
+
+    time.sleep(0.05)
```

Comment:
Started with this way from examples I found online.
I am switching to polling with a timeout in the next version so a fixed delay can be avoided.